### PR TITLE
Documented connection hints as well as keep-alives within all protocol states

### PR DESCRIPTION
**Pull request by dotStart**

### CODE REVIEW

File: bolt/bolt-protocol-message-specification-4.md [hunk 1]
```diff
@@ -1249,8 +1249,117 @@ FAILURE {"code": "Example.Failure.Code", "message": "example failure"}
...
+The changes compared to Bolt protocol version 4.2 are listed below:
+
+ * `NOOP` chunks may now be transmitted in all connection states when a connection remains in idle for extended
+   periods of time while the server is busy processing a request.
+ * An additional `hint` dictionary has been added to the `metadata` property of the `SUCCESS` structure transmitted in
+   response to the `HELLO` command in order to provide optional configuration hints to drivers.
+
+### Request Message - 4.3 - `HELLO`
+
+The `HELLO` message request the connection to be authorized for use with the remote database.
```
Comment (robsdedude) [inline bolt/bolt-protocol-message-specification-4.md]: request*s*?

File: bolt/bolt-protocol-message-specification-4.md [hunk 2]
```diff
@@ -1249,8 +1249,117 @@ FAILURE {"code": "Example.Failure.Code", "message": "example failure"}
...
+#### Server Response `SUCCESS`
+
+A `SUCCESS` message response indicates that the client is permitted to exchange further messages.
+Servers can include metadata that describes details of the server environment and/or the connection.
+
+The following fields are defined for inclusion in the `SUCCESS` metadata.
+
+- `server::String` (server agent string, example `"Neo4j/4.1.0"`)
+- `connection_id::String` (unique identifier of the bolt connection used on the server side, example: `"bolt-61"`)
+- `hints::Dictionary` (set of optional configuration hints to be considered by the driver )
```
Comment (robsdedude) [inline bolt/bolt-protocol-message-specification-4.md]: space before `)`

File: bolt/bolt-protocol-message-specification-4.md [hunk 3]
```diff
@@ -1249,8 +1249,117 @@ FAILURE {"code": "Example.Failure.Code", "message": "example failure"}
...
+
+The new `hints` dictionary may contain a set of optional configuration hints which may be interpreted or ignored by
+drivers at their own discretion in order to augment operations where applicable. A full listing of the available
+hints may be found in [Appendix B](#appendix-b---connection-hints).
+
+Example:
+
+```
+SUCCESS {"server": "Neo4j/4.0.0", "hints": {"connection.recv_timeout_seconds":120}}
```
Comment (robsdedude) [inline bolt/bolt-protocol-message-specification-4.md]: `{"connection.recv_timeout_seconds": 120}`
(added space)

**Review by robsdedude — COMMENTED** [review]
Looks good except for minor formatting.

File: bolt/bolt-protocol-message-specification-4.md [hunk 4]
```diff
@@ -1249,8 +1249,119 @@ FAILURE {"code": "Example.Failure.Code", "message": "example failure"}
...
 
+# Version 4.3
+
+## Deltas
+
+The changes compared to Bolt protocol version 4.2 are listed below:
+
+ * `NOOP` chunks may now be transmitted in all connection states when a connection remains in idle for extended
+   periods of time while the server is busy processing a request.
+ * An additional `hint` dictionary has been added to the `metadata` property of the `SUCCESS` structure transmitted in
```
Comment (fbiville) [inline bolt/bolt-protocol-message-specification-4.md]:
```suggestion
 * An additional `hints` dictionary has been added to the `metadata` property of the `SUCCESS` structure transmitted in
```

File: bolt/bolt-protocol-message-specification-4.md [hunk 5]
```diff
@@ -1323,3 +1434,24 @@ S: SUCCESS {"type": "r", "db": "test"}
...
+
+### Appendix B - Connection Hints
+
+As of version 4.3, the server may optionally include a set of configuration hints within the `SUCCESS` metadata provided
+in response to the `HELLO` message. Drivers may choose to interpret or ignore these hints at their own discretion. The
+following set of hints are currently defined:
+
+#### connection.recv_timeout_seconds - 4.3
+
+Maximum amount of time (in seconds) for which a connection may remain in idle following a request before drivers should
```
Comment (fbiville) [inline bolt/bolt-protocol-message-specification-4.md]:
```suggestion
Maximum amount of time (in seconds) for which a connection may remain idle following a request before drivers should
```

Comment (fbiville) [reply bolt/bolt-protocol-message-specification-4.md]: might be worth simplifying the sentence a bit

File: bolt/bolt-protocol-message-specification-4.md [hunk 6]
```diff
@@ -1323,3 +1434,24 @@ S: SUCCESS {"type": "r", "db": "test"}
...
+consider it stale.
+
+The server will ensure that a chunk (or `NOOP` chunk) will be transmitted at least once per timeout period. Drivers
```
Comment (fbiville) [inline bolt/bolt-protocol-message-specification-4.md:1449]:
```suggestion
The server will ensure that a chunk (like a `NOOP` chunk e.g.) will be transmitted at least once per timeout period. Drivers
```

File: bolt/bolt-protocol-message-specification-4.md [hunk 7]
```diff
@@ -1323,3 +1434,24 @@ S: SUCCESS {"type": "r", "db": "test"}
...
+which elect to interpret this hint may terminate connections which are considered stale by the definition of this hint.
```
Comment (fbiville) [inline bolt/bolt-protocol-message-specification-4.md]:
```suggestion
which choose to interpret this hint may terminate connections which are considered stale by the definition of this hint.
```

**Review by fbiville — APPROVED** [review]

**Pull request merged**